In [21]:
# Import necessary libraries
#!pip install ISLP
from ISLP import load_data
from ISLP.models import (ModelSpec as MS,
                         summarize,
                         poly)
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.stats.anova import anova_lm
import matplotlib.pyplot as plt

In [ ]:
#1 Load the College dataset using the ISLP package. Display the first 10 rows and list all column names. Briefly describe the dataset and the meaning of the variables Outstate, Private, Room.Board, PhD, and Top10perc.

data = load_data('College')
columns = list(data.columns)
columns


['Private',
 'Apps',
 'Accept',
 'Enroll',
 'Top10perc',
 'Top25perc',
 'F.Undergrad',
 'P.Undergrad',
 'Outstate',
 'Room.Board',
 'Books',
 'Personal',
 'PhD',
 'Terminal',
 'S.F.Ratio',
 'perc.alumni',
 'Expend',
 'Grad.Rate']

In [ ]:
data.head(10)

Apps        Accept       Enroll   Top10perc   Top25perc  \
count    777.000000    777.000000   777.000000  777.000000  777.000000   
mean    3001.638353   2018.804376   779.972973   27.558559   55.796654   
std     3870.201484   2451.113971   929.176190   17.640364   19.804778   
min       81.000000     72.000000    35.000000    1.000000    9.000000   
25%      776.000000    604.000000   242.000000   15.000000   41.000000   
50%     1558.000000   1110.000000   434.000000   23.000000   54.000000   
75%     3624.000000   2424.000000   902.000000   35.000000   69.000000   
max    48094.000000  26330.000000  6392.000000   96.000000  100.000000   

        F.Undergrad   P.Undergrad      Outstate   Room.Board        Books  \
count    777.000000    777.000000    777.000000   777.000000   777.000000   
mean    3699.907336    855.298584  10440.669241  4357.526384   549.380952   
std     4850.420531   1522.431887   4023.016484  1096.696416   165.105360   
min      139.000000      1.000000   2340.000000  1780.000000    96.000000   
25%      992.000000     95.000000   7320.000000  3597.000000   470.000000   
50%     1707.000000    353.000000   9990.000000  4200.000000   500.000000   
75%     4005.000000    967.000000  12925.000000  5050.000000   600.000000   
max    31643.000000  21836.000000  21700.000000  8124.000000  2340.000000   

          Personal         PhD    Terminal   S.F.Ratio  perc.alumni  \
count   777.000000  777.000000  777.000000  777.000000   777.000000   
mean   1340.642214   72.660232   79.702703   14.089704    22.743887   
std     677.071454   16.328155   14.722359    3.958349    12.391801   
min     250.000000    8.000000   24.000000    2.500000     0.000000   
25%     850.000000   62.000000   71.000000   11.500000    13.000000   
50%    1200.000000   75.000000   82.000000   13.600000    21.000000   
75%    1700.000000   85.000000   92.000000   16.500000    31.000000   
max    6800.000000  103.000000  100.000000   39.800000    64.000000   

             Expend  Grad.Rate  
count    777.000000  777.00000  
mean    9660.171171   65.46332  
std     5221.768440   17.17771  
min     3186.000000   10.00000  
25%     6751.000000   53.00000  
50%     8377.000000   65.00000  
75%    10830.000000   78.00000  
max    56233.000000  118.00000

In [14]:
data.describe()

Apps        Accept       Enroll   Top10perc   Top25perc  \
count    777.000000    777.000000   777.000000  777.000000  777.000000   
mean    3001.638353   2018.804376   779.972973   27.558559   55.796654   
std     3870.201484   2451.113971   929.176190   17.640364   19.804778   
min       81.000000     72.000000    35.000000    1.000000    9.000000   
25%      776.000000    604.000000   242.000000   15.000000   41.000000   
50%     1558.000000   1110.000000   434.000000   23.000000   54.000000   
75%     3624.000000   2424.000000   902.000000   35.000000   69.000000   
max    48094.000000  26330.000000  6392.000000   96.000000  100.000000   

        F.Undergrad   P.Undergrad      Outstate   Room.Board        Books  \
count    777.000000    777.000000    777.000000   777.000000   777.000000   
mean    3699.907336    855.298584  10440.669241  4357.526384   549.380952   
std     4850.420531   1522.431887   4023.016484  1096.696416   165.105360   
min      139.000000      1.000000   2340.000000  1780.000000    96.000000   
25%      992.000000     95.000000   7320.000000  3597.000000   470.000000   
50%     1707.000000    353.000000   9990.000000  4200.000000   500.000000   
75%     4005.000000    967.000000  12925.000000  5050.000000   600.000000   
max    31643.000000  21836.000000  21700.000000  8124.000000  2340.000000   

          Personal         PhD    Terminal   S.F.Ratio  perc.alumni  \
count   777.000000  777.000000  777.000000  777.000000   777.000000   
mean   1340.642214   72.660232   79.702703   14.089704    22.743887   
std     677.071454   16.328155   14.722359    3.958349    12.391801   
min     250.000000    8.000000   24.000000    2.500000     0.000000   
25%     850.000000   62.000000   71.000000   11.500000    13.000000   
50%    1200.000000   75.000000   82.000000   13.600000    21.000000   
75%    1700.000000   85.000000   92.000000   16.500000    31.000000   
max    6800.000000  103.000000  100.000000   39.800000    64.000000   

             Expend  Grad.Rate  
count    777.000000  777.00000  
mean    9660.171171   65.46332  
std     5221.768440   17.17771  
min     3186.000000   10.00000  
25%     6751.000000   53.00000  
50%     8377.000000   65.00000  
75%    10830.000000   78.00000  
max    56233.000000  118.00000

The dataset "College" contains statistics for 777 US Colleges from the 1995 issue of US News and World Report. The variables of interest are:
1. Outstate: This variable represents the out-of-state tuition fee.
2. Private: A factor with levels No and Yes indicating private or public university.
3. Room.Board: This variable represents the room and board costs.
4. PhD: This variable indicates the percentage of faculty with PhDs at the institution.
5. Top10perc: This variable represents the percentage of students from the top 10% of their high school class.

In [ ]:
#2 Fit a simple linear regression model predicting Outstate tuition using Top10perc

X = pd.DataFrame({'intercept': np.ones(data.shape[0]),
                  'Top10perc': data['Top10perc']})
y = data['Outstate']

model = sm.OLS(y, X)
results2 = model.fit()
summarize(results)

coef  std err       t  P>|t|
intercept  6906.4586  221.614  31.164    0.0
Top10perc   128.2437    6.774  18.931    0.0

The coefificient for Top10perc is 128.2437. This indicates that for each additional percentage point increase in the proportion of students from the top 10% of their high school class, the out-of-state tuition fee is expected to increase by approximately \$128.24, assuming all other factors remain constant. This suggests a positive relationship between the academic caliber of incoming students and the tuition fees charged by the institution.

In [22]:
#3 Fit a multiple linear regression model predicting Outstate tuition using Top10perc, Room.Board, and PhD.
X = pd.DataFrame({'intercept': np.ones(data.shape[0]),
                  'Top10perc': data['Top10perc'],
                  'Room.Board': data['Room.Board'],
                  'PhD': data['PhD']})
y = data['Outstate']
model = sm.OLS(y, X)
results3 = model.fit()
summarize(results)

coef  std err       t  P>|t|
intercept         -1087.2758  573.149  -1.897  0.058
Top10perc           138.0675   18.661   7.399  0.000
Room.Board            1.9081    0.097  19.712  0.000
PhD                   1.9617    7.194   0.273  0.785
Top10perc_squared    -0.6859    0.213  -3.218  0.001

The coefficients for Top10perc is 82.0003, Room.Board is 1.8825, and the PhD is 5.6226.
That indicates that for each additional percentage point increase in the proportion of students from the top 10% of their high school class, the out-of-state tuition fee is expected to increase by approximately $82.00, assuming all other factors remain constant. For each additional dollar increase in room and board costs, the out-of-state tuition fee is expected to increase by approximately \$1.88, assuming all other factors remain constant. Lastly, for each additional percentage point increase in the proportion of faculty with PhDs, the out-of-state tuition fee is expected to increase by approximately \$5.62, assuming all other factors remain constant. However, the coefficient for PhD is insignificant.
This suggests that higher academic standards and better facilities are associated with higher tuition fees.

In [23]:
#4 Add a quadratic term for Top10perc to the previous model.
X = pd.DataFrame({'intercept': np.ones(data.shape[0]),
                  'Top10perc': data['Top10perc'],
                  'Room.Board': data['Room.Board'],
                  'PhD': data['PhD'],
                  'Top10perc_squared': data['Top10perc'] ** 2})
y = data['Outstate']
model = sm.OLS(y, X)
results4 = model.fit()
summarize(results)

coef  std err       t  P>|t|
intercept         -1087.2758  573.149  -1.897  0.058
Top10perc           138.0675   18.661   7.399  0.000
Room.Board            1.9081    0.097  19.712  0.000
PhD                   1.9617    7.194   0.273  0.785
Top10perc_squared    -0.6859    0.213  -3.218  0.001

The coefficient for Top10perc is 138.0675, Room.Board is 1.9081, PhD is 1.9671, and Top10perc_squared is -0.6859.
The coefficient for Top10perc_squared indicates a negative quadratic relationship between Top10perc and Outstate tuition. This suggests that while an increase in the percentage of students from the top 10% of their high school class is associated with higher out-of-state tuition fees, the rate of increase in tuition fees diminishes as Top10perc continues to rise. In other words, the impact of Top10perc on tuition is more pronounced at lower levels of Top10perc and becomes less significant at higher levels, indicating a concave downward relationship.

In [24]:
#5 Use anova_lm to compare the linear and quadratic models from questions 3 and 4. Report the F-statistic and p-value, and state whether the quadratic term significantly improves the model.

anova_lm(results3, results4)

df_resid           ssr  df_diff       ss_diff          F    Pr(>F)
0     773.0  5.693394e+09      0.0           NaN        NaN       NaN
1     772.0  5.618020e+09      1.0  7.537388e+07  10.357499  0.001343

The quadratic term significantly improves the model, as indicated by the F-statistic of 10.357 and a p-value of 0.001. This low p-value suggests that the addition of the quadratic term for Top10perc provides a statistically significant improvement in explaining the variance in Outstate tuition compared to the linear model alone.